Documentation for Task Group Dynamic Dependencies (#1863)
uxlfoundation/oneTBB: doc/main/reference/task_group_bypass_support.rst

.. _task_group_bypass_support:

Task Bypass Support for ``task_group``
======================================

.. note::
    To enable this extension, define the ``TBB_PREVIEW_TASK_GROUP_EXTENSIONS`` macro with a value of ``1``.

.. contents::
    :local:
    :depth: 2

Description
***********

The |full_name| implementation extends the requirements for user-provided function object from
`tbb::task_group specification <https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onetbb/source/task_scheduler/task_group/task_group_cls>`_

> to allow them to return a ``task_handle`` object.

Reviewer: It seems that [task_group specification](https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onetbb/source/task_scheduler/task_group/task_group_cls) is silent on the return type of the user-provided function object.

`Task Bypassing <../tbb_userguide/Task_Scheduler_Bypass.html>`_ allows developers to reduce task scheduling overhead by providing a hint about
which task should be executed next.

> Execution of the hinted task is not guaranteed to occur immediately, nor to be performed by the same thread.

Reviewer: ```suggestion
Execution of the deferred task owned by a returned ``task_handle`` is not guaranteed to occur immediately, nor to be performed by the same thread.
```

> Execution of the deferred task owned by a returned ``task_handle`` is not guaranteed to occur immediately, nor to be performed by the same thread.

Author: I have added a note at the end of description.

            public:
                // Only the requirements for the return type of function F are changed
                template <typename F>
                task_handle defer(F&& f);

                // Only the requirements for the return type of function F are changed
                template <typename F>
                task_group_status run_and_wait(const F& f);

                // Only the requirements for the return type of function F are changed
                template <typename F>
                void run(F&& f);
            }; // class task_group
        } // namespace tbb
    } // namespace oneapi

Member Functions
----------------

.. code:: cpp

    template <typename F>
    task_handle defer(F&& f);

    template <typename F>
    task_group_status run_and_wait(const F& f);

    template <typename F>
    void run(F&& f);

The function object ``F`` may return a ``task_handle`` object. If the returned handle is non-empty and owns a task without dependencies, it serves as an optimization
hint for a task that could be executed next.

If the returned handle was created by a ``task_group`` other than ``*this``, the behavior is undefined.

Example
-------

> The example below demonstrates how to implement a parallel for loop using ``task_group`` and divide-and-conquer pattern.

Reviewer 2: ```suggestion
The example below demonstrates how to process a sequence in parallel using ``task_group`` and the divide-and-conquer pattern.
```
